### PR TITLE
Store when a gateway is disconnected

### PR DESCRIPTION
**Pull request by nicholaspcr**
#### Summary

This PR allows the registering of the last time the connection with a gateway ended, being saved in a field called `disconnected_at`.

Image of the connStats object in the redis store:
image

References #4850
...

#### Changes

- Add the field `disconnected_at` to the GatewayConnectionStats definition
- Change GS's registry interface, a TTL parameter in the Set method

#### Testing

Manual testing

##### Regressions

...

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [x] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: api/gateway.proto [hunk 1]
```diff
@@ -391,4 +391,6 @@ message GatewayConnectionStats {
   }
   // Statistics for each sub band.
   repeated SubBand sub_bands = 10;
+
+  google.protobuf.Timestamp disconnected_at = 11;
```
Comment (adriansmares) [inline api/gateway.proto]: I would keep the field below `connected_at`.

File: cmd/internal/shared/gatewayserver/config.go [hunk 2]
```diff
@@ -66,4 +66,6 @@ var DefaultGatewayServerConfig = gatewayserver.Config{
 		Listen:                 ":1887",
 		ListenTLS:              ":8887",
 	},
+
+	DisconnectTTL: 48 * time.Hour,
```
Comment (adriansmares) [inline cmd/internal/shared/gatewayserver/config.go]: No new line, and also probably sits better under `UpdateConnectionStatsDebounceTime`.

File: pkg/gatewayserver/config.go [hunk 3]
```diff
@@ -66,6 +66,8 @@ type Config struct {
 	MQTTV2       config.MQTT        `name:"mqtt-v2"`
 	UDP          UDPConfig          `name:"udp"`
 	BasicStation BasicStationConfig `name:"basic-station"`
+
+	DisconnectTTL time.Duration `name:"disconnect-ttl"`
```
Comment (adriansmares) [inline pkg/gatewayserver/config.go]: I would keep this under `UpdateConnectionStatsDebounceTime`, and I think a better name would be `ConnectionStatsDisconnectTTL`. You should also add a `description`.

File: pkg/gatewayserver/redis/registry_test.go [hunk 4]
```diff
@@ -153,7 +153,7 @@ func TestRegistry(t *testing.T) {
 			"uplink_count",
 			"last_uplink_received_at",
 			"downlink_count",
-		})
+		}, 0)
 		a.So(err, should.BeNil)
 		retrieved, err = registry.Get(ctx, ids)
 		a.So(err, should.BeNil)
```
Comment (adriansmares) [inline pkg/gatewayserver/redis/registry_test.go:181]: You probably can add a TTL test here (stored disconnected at with TTL, expect the immediate `Get` to work, sleep the TTL, check that `Get` returns not found).

**Review by adriansmares — COMMENTED** [review]
Looks good functionally, only a couple of style / naming nitpicks.

Please rebase your commits such that they follow the style in `CONTRIBUTING.md` - commits such as 8f0feeb9613b0cff970822e1cb64eb4a782a9dbd and 04e7a90377bda21cf658925b0d27b20fece8c547 can be squashed, and should have the `api` prefix. The commit message should also be written using the imperative tense.

Also, when you regenerate the protocol buffers, it may be a good idea to also do `jssdk:definitions` - sometimes these change too after updating the definitions.

File: pkg/gatewayserver/redis/registry_test.go [hunk 5]
```diff
@@ -109,8 +109,30 @@ func TestRegistry(t *testing.T) {
...
+		a.So(registry.Set(ctx, ids, nil, nil, 0), should.BeNil)
+	})
+
+	t.Run("SetWithTTL", func(t *testing.T) {
+		a, ctx := test.New(t)
+		stats := &ttnpb.GatewayConnectionStats{
+			DisconnectedAt: ttnpb.ProtoTimePtr(time.Date(2021, 12, 02, 11, 24, 58, 0, time.UTC)),
+		}
+
+		err := registry.Set(ctx, ids, stats, []string{"disconnected_at"}, 2*time.Second)
```
Comment (adriansmares) [inline pkg/gatewayserver/redis/registry_test.go]: You can use smaller values, and checkout our usage of `test.Delay`. That one scales with the `TEST_SLOWDOWN` environment variable (we generally use 8).  Also see the `Timeout` variable in this file.

**Review by adriansmares — APPROVED** [review]
Good stuff !

File: pkg/gatewayserver/redis/registry_test.go (hunk 5 above)
Comment (nicholaspcr) [reply pkg/gatewayserver/redis/registry_test.go]: I didn't even notice it, it's a cool idea. Changed the test to use it.

File: api/gateway.proto [hunk 6]
```diff
@@ -365,6 +365,7 @@ message GatewayStatus {
 // Connection stats as monitored by the Gateway Server.
 message GatewayConnectionStats {
   google.protobuf.Timestamp connected_at = 1;
+  google.protobuf.Timestamp disconnected_at = 11;
```
Comment (htdvisser) [inline api/gateway.proto:368]: We usually write a `// next: 12` at the bottom of the message when we go out-of-order.

**Review by htdvisser — APPROVED** [review]

**Comment by htdvisser** [comment]
I forgot to add here:

1. Please write a CHANGELOG entry with details on the API and Config changes
2. Please sync with our frontend devs so that the Console displays this correctly (instead of showing disconnected gateways as connected)

**Pull request merged**